pkg/spoke/managedclusteraddon/lease_controller.go

```diff
@@ -0,0 +1,182 @@
+package managedclusteraddon
+
+import (
+	"context"
+	"time"
+
+	addonapiv1alpha1 "github.com/open-cluster-management/api/addon/v1alpha1"
+	addonclientv1alpha1 "github.com/open-cluster-management/api/client/addon/clientset/versioned"
+	addoninformerv1alpha1 "github.com/open-cluster-management/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "github.com/open-cluster-management/api/client/addon/listers/addon/v1alpha1"
+	"github.com/open-cluster-management/registration/pkg/helpers"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	coordinformers "k8s.io/client-go/informers/coordination/v1"
+	coordlisters "k8s.io/client-go/listers/coordination/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const leaseDurationTimes = 5
+
+type managedClusterAddOnLeaseController struct {
+	clusterName string
+	addOnClient addonclientv1alpha1.Interface
+	addOnLister addonlisterv1alpha1.ManagedClusterAddOnLister
+	leaseLister coordlisters.LeaseLister
+}
+
+func NewManagedClusterAddOnLeaseController(clusterName string,
+	addOnClient addonclientv1alpha1.Interface,
+	addOnInformer addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	leaseInformer coordinformers.LeaseInformer,
+	resyncInterval time.Duration,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterAddOnLeaseController{
+		clusterName: clusterName,
+		addOnClient: addOnClient,
+		addOnLister: addOnInformer.Lister(),
+		leaseLister: leaseInformer.Lister(),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(c.queueKeyFunc, leaseInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(resyncInterval).
+		ToController("ManagedClusterAddOnLeaseController", recorder)
+}
+
+func (c *managedClusterAddOnLeaseController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	if queueKey == factory.DefaultQueueKey {
+		addOns, err := c.addOnLister.ManagedClusterAddOns(c.clusterName).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		errs := []error{}
+		for _, addOn := range addOns {
+			addOnInstallNamespace, ok := addOn.Annotations["installNamespace"]
+			if !ok {
+				// addon does not have install namespace, ignore it.
+				continue
+			}
+			if err := c.syncSingle(ctx, addOnInstallNamespace, addOn, syncCtx.Recorder()); err != nil {
```

Comment:
we need some doc about the format of the lease we accept.

Reply:
I think we can require that the addon lease name must be same with the addon name, what's your opinion? or we use a specific format, like <addon-name>-lease?